File: python-algo/leetcode/problem_513.py
"""513.
"""
from typing import Optional
from common.tree_node import TreeNode


class Solution:
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        if not root:
            return None
        q = [root]
        while q:
            is_end = True
            n = len(q)
            leftmost = q[0]
            for i in range(n):
                node = q.pop(0)
                if node.left or node.right:
                    is_end = False
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            if is_end:
                return leftmost.val
        return None

File: java-algo/src/main/java/com/leetcode/Problem513.java
package com.leetcode;

import com.common.TreeNode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * 513. Find Bottom Left Tree Value
 * https://leetcode.com/problems/find-bottom-left-tree-value/
 */
public class Problem513 {
    public int findBottomLeftValue(TreeNode root) {
        Deque<TreeNode> q = new ArrayDeque<>();
        q.push(root);
        while (!q.isEmpty()) {
            TreeNode first = q.getFirst();
            boolean hasChild = false;
            int n = q.size();
            for (int i = 0; i < n; i++) {
                TreeNode node = q.pollFirst();
                if (node.left != null) {
                    hasChild = true;
                    q.offerLast(node.left);
                }
                if (node.right != null) {
                    hasChild = true;
                    q.offerLast(node.right);
                }
            }
            if (!hasChild) {
                return first.val;
            }
        }
        return 0;
    }
}
